File: frontend/src/types/village.ts
export interface VillageChannel {
  id: string;
  name: string;
  slug: string;
  description: string;
  stage_id?: number | null;
  is_private: boolean;
  posts_count: number;
}

export interface PaginatedVillageChannels {
  items: VillageChannel[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface VillagePost {
  id: string;
  channel_id: string;
  channel_name?: string | null;
  channel_slug?: string | null;
  user_id: string;
  author_alias: string;
  author_avatar_seed: string;
  title: string;
  content: string;
  status: string;
  is_flagged: boolean;
  upvotes_count: number;
  comments_count: number;
  has_reacted: boolean;
  user_reaction?: string | null;
  created_at: string;
  updated_at: string;
}

export interface VillagePostCreateRequest {
  channel_id: string;
  title: string;
  content: string;
  author_alias?: string;
  author_avatar_seed?: string;
}

export interface VillagePostUpdateRequest {
  title?: string;
  content?: string;
}

export interface PaginatedVillagePosts {
  items: VillagePost[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface VillageComment {
  id: string;
  post_id: string;
  user_id: string;
  author_alias: string;
  content: string;
  status: string;
  created_at: string;
}

export interface VillageCommentCreateRequest {
  content: string;
  author_alias?: string;
}

export interface VillageCommentUpdateRequest {
  content: string;
}

export interface PaginatedVillageComments {
  items: VillageComment[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface VillageReactionRequest {
  reaction_type: 'heart' | 'hug' | 'celebrate' | 'strength' | 'helpful' | string;
}

export interface VillageReactionResponse {
  post_id: string;
  reaction_type: string;
  action: 'added' | 'removed' | string;
  upvotes_count: number;
  has_reacted: boolean;
}

export interface VillageReportCreateRequest {
  reason: 'harassment' | 'medical_misinformation' | 'hate_or_abuse' | 'spam' | 'inappropriate_content' | 'privacy_violation' | 'other' | string;
  details?: string;
}

export interface VillageReportResponse {
  id: string;
  post_id?: string | null;
  comment_id?: string | null;
  reason: string;
  details?: string | null;
  status: string;
  created_at: string;
}

File: backend/app/schemas/village.py
from datetime import datetime
from typing import Any, Dict, List, Optional
import uuid
from pydantic import BaseModel, ConfigDict, Field, field_validator


# ============================================================================
# Channel Schemas
# ============================================================================

class VillageChannelResponse(BaseModel):
    id: uuid.UUID
    name: str
    slug: str
    description: str
    stage_id: Optional[int] = None
    is_private: bool = False
    posts_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class PaginatedVillageChannelsResponse(BaseModel):
    items: List[VillageChannelResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
    has_next: bool
    has_prev: bool


# ============================================================================
# Post Schemas
# ============================================================================

class VillagePostCreateRequest(BaseModel):
    channel_id: uuid.UUID
    title: str = Field(..., min_length=3, max_length=255, description="Post title")
    content: str = Field(..., min_length=5, max_length=10000, description="Post content text")
    author_alias: Optional[str] = Field(default=None, max_length=100, description="Display alias (defaults to user name)")
    author_avatar_seed: Optional[str] = Field(default="avatar1", max_length=100)

    @field_validator("title", mode="before")
    @classmethod
    def validate_title(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("Title cannot be empty.")
        return v.strip()

    @field_validator("content", mode="before")
    @classmethod
    def validate_content(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("Content cannot be empty.")
        return v.strip()


class VillagePostUpdateRequest(BaseModel):
    title: Optional[str] = Field(default=None, min_length=3, max_length=255)
    content: Optional[str] = Field(default=None, min_length=5, max_length=10000)


class VillagePostResponse(BaseModel):
    id: uuid.UUID
    channel_id: uuid.UUID
    channel_name: Optional[str] = None
    channel_slug: Optional[str] = None
    user_id: uuid.UUID
    author_alias: str
    author_avatar_seed: str
    title: str
    content: str
    status: str
    is_flagged: bool
    upvotes_count: int
    comments_count: int
    has_reacted: bool = False
    user_reaction: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaginatedVillagePostsResponse(BaseModel):
    items: List[VillagePostResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
    has_next: bool
    has_prev: bool


# ============================================================================
# Comment Schemas
# ============================================================================

class VillageCommentCreateRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=3000, description="Comment text")
    author_alias: Optional[str] = Field(default=None, max_length=100)


class VillageCommentUpdateRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=3000, description="Updated comment text")


class VillageCommentResponse(BaseModel):
    id: uuid.UUID
    post_id: uuid.UUID
    user_id: uuid.UUID
    author_alias: str
    content: str
    status: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaginatedVillageCommentsResponse(BaseModel):
    items: List[VillageCommentResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
    has_next: bool
    has_prev: bool


# ============================================================================
# Reaction Schemas
# ============================================================================

VALID_REACTION_TYPES = {"heart", "hug", "celebrate", "strength", "helpful"}

class VillageReactionRequest(BaseModel):
    reaction_type: str = Field(default="heart", description="Type of reaction: heart, hug, celebrate, strength, helpful")


class VillageReactionResponse(BaseModel):
    post_id: uuid.UUID
    reaction_type: str
    action: str  # "added" or "removed"
    upvotes_count: int
    has_reacted: bool


# ============================================================================
# Report & Moderation Schemas
# ============================================================================

VALID_REPORT_REASONS = {
    "harassment",
    "medical_misinformation",
    "hate_or_abuse",
    "spam",
    "inappropriate_content",
    "privacy_violation",
    "other",
}

class VillageReportCreateRequest(BaseModel):
    reason: str = Field(..., description="Reporting reason category")
    details: Optional[str] = Field(default=None, max_length=2000, description="Optional description of the issue")


class VillageReportResponse(BaseModel):
    id: uuid.UUID
    post_id: Optional[uuid.UUID] = None
    comment_id: Optional[uuid.UUID] = None
    reason: str
    details: Optional[str] = None
    status: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class VillageModerationActionRequest(BaseModel):
    action: str = Field(..., description="Action to take: 'dismiss', 'hide_content', 'resolve'")
    note: Optional[str] = Field(default=None, max_length=1000)


class PaginatedVillageReportsResponse(BaseModel):
    items: List[VillageReportResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
    has_next: bool
    has_prev: bool
